lib/utmp.c

```diff
@@ -246,7 +246,7 @@ static
 #ifdef HAVE_STRUCT_UTMP_UT_HOST
 	} else if (   (NULL != ut)
 	           && ('\0' != ut->ut_host[0])) {
-		hostname = XMALLOC(SIZEOF_ARRAY(ut->ut_host) + 1, char);
+		hostname = XMALLOC(NITEMS(ut->ut_host) + 1, char);
```

Comment:
This doesn't seem right.  If ut_host were to become something other than char, the result would be wrong, whereas with the current use of SIZEOF_ARRAY it would continue to work.  So auto-detecting the base type, to pass as second arg to XMALLOC, would be ok.

Reply:
If ut_host would become something other than char, then `hostname` should also become something other than char.

---
Regarding auto-detecting the base type:

The second argument to [X]MALLOC() is coupled with the asignee, thanks to the cast of the return type:

```c
$ grepc XMALLOC .
./lib/alloc.h:#define XMALLOC(n, type)  ((type *) xmallocarray(n, sizeof(type)))
```

That cast ensures that if `hostname` changed to `wchar_t`, the compiler would complain about implicit conversion from `char *` to `wchar_t *` at the assignment.

---

Regarding SIZEOF_ARRAY() vs NITEMS():

[X]MALLOC()'s arguments are 1) number of elements 2) type of each element.  Then, the [x]mallocarray() function performs the multiplication nelem * sizeof_elem, to pass the size to the actual malloc(3).

NITEMS() returns a number of elements, which is why it's the correct call.  SIZEOF_ARRAY() returns total size of the array, that is nelem * sizeof_elem.  If we pass SIZEOF_ARRAY() to MALLOC(), mallocarray() will multiply again by sizeof_elem, resulting in nelem * sizeof_elem * sizeof_elem.  That's conceptually wrong, even if sizeof_elem happens to be 1 in the present case.

Reply:
Let me show an example:


```c
wchar_t  src[4] = L"foo";
char     *c;
wchar_t  *w;

c = MALLOC(4, wchar_t); // Warning: implicit conversion for char* to wchar_t*
p = MALLOC(SIZEOF_ARRAY(src), wchar_t);  // We're allocating too much!  4 * sizeof(wchar_t) * sizeof(wchar_t)
p = MALLOC(NITEMS(src), wchar_t);  // Ok

Reply:
yeah but too much is ok, too little is not :)

but the tying to dest type is enough for me, thanks.

Reply:
> yeah but too much is ok, too little is not :)

Except when that may cause confusion.  :)

> but the tying to dest type is enough for me, thanks.

Good!